docs/src/diplomacy/adder_tutorial.md

```diff
@@ -0,0 +1,327 @@
+a Chisel `UInt` with the width described by our edge parameter (see the `bundle`
+function).
+
+```scala mdoc
+// PARAMETER TYPES:                       D              U            E          B
+object AdderNodeImp extends SimpleNodeImp[DownwardParam, UpwardParam, EdgeParam, UInt] {
+  def edge(pd: DownwardParam, pu: UpwardParam, p: Parameters, sourceInfo: SourceInfo) = {
+    if (pd.width < pu.width) EdgeParam(pd.width) else EdgeParam(pu.width)
+  }
+  def bundle(e: EdgeParam) = UInt(e.width.W)
+  def render(e: EdgeParam) = RenderedEdge("blue", s"width = ${e.width}")
+}
+```
+
+Notice that the `edge` function does the actual negotiation between nodes. In our
+case, it compares the parameters traveling both upwards and downwards, and chooses
+the one with the smaller width as the final result to make visible from each node.
+
+The `render` function is required by `NodeImp`s and holds metadata for rendering
+a view of the negotiated information in some graphical format.
+
+#### Nodes
+
+Nodes can receive or generate parameters. A module parameterized through
+Diplomacy must have one or more nodes in order to receive or send parameters.
+Here, we will create some different types of nodes for use in the modules we
+would like to parameterize.
+
+Let's start off with creating a node for our driver. We will make it a
+`SourceNode` with the node implementation shown in the previous section, since
+`SourceNode`s only generate downward-flowing parameters along outward edges.
+
+For our `AdderDriverNode`, `widths` of type `Seq[DownwardParam]` represents the
+desired widths for the output wires of the module instantiating this node
+(`AdderDriver`). We are using a `Seq` here because each node can send multiple
+outputs. In our example, the `AdderDriverNode` will send two identical outputs:
+one to the adder, and the other to the monitor. These properties are `require`d
+by the node.
+
+```scala mdoc
+/** node for [[AdderDriver]] (source) */
+class AdderDriverNode(widths: Seq[DownwardParam])(implicit valName: ValName)
+  extends SourceNode(AdderNodeImp)(widths) {
+  require(widths.length == 2)
+  require(widths.head.width == widths.tail.head.width)
+}
+```
+
+Our monitor node follows the same pattern, this time extending `SinkNode`, a
+type of node that only generates upward-flowing parameters along inward edges.
+The parameter it takes in is `width` of type `UpwardParam`.
+
+```scala mdoc
+/** node for [[AdderMonitor]] (sink) */
+class AdderMonitorNode(width: UpwardParam)(implicit valName: ValName)
+  extends SinkNode(AdderNodeImp)(Seq(width))
+```
+
+Our adder node will take inputs from two `AdderDriverNode`s (the two addends)
+and pass one output (the sum) into the monitor. Because the number of inputs and
+outputs differ, we will make this node extend a `NexusNode`.
+
+`dFn` maps **incoming downward** parameters into **outgoing downward** parameters.
+Similarly, `uFn` maps **incoming upward** parameters into **outgoing upward** ones.
+
```

Comment:
Can you clarify what an "incoming" and an "outgoing" parameter is?

Reply:
Added some clarification here: https://github.com/chipsalliance/rocket-chip/pull/2615/commits/4dbe86e7af38246bb928d4e0b218dabc8883b865

Hope that helps!

Reply:
`uFn` maps **upward** parameters along **inward edges** into **upward** parameters along **outward edges**.`
=>
`uFn` maps **upward** parameters along **outward edges** into **upward** parameters along **inward edges**.`